src/app.rs

```diff
@@ -307,35 +340,25 @@ pub fn load_config<P: AsRef<Path>>(path: P) -> io::Result<Config> {
     Ok(config_data)
 }
 
-#[derive(Clone)]
-pub struct AppState {
-    pub db: Addr<DbExecutor>,
-    pub config: Arc<Config>,
-    pub job_queue: Addr<JobQueue>,
-    pub delta_generator: Addr<DeltaGenerator>,
-}
+fn handle_build_repo(config: Data<Config>,
+                     req: HttpRequest) -> Result<HttpResponse, actix_web::Error> {
+    let tail = req.match_info().query("tail");
+    let id = req.match_info().query("id");
```

Comment:
@alexlarsson by id i meant this

i don't thinks it's checked against any config
and even if it can't contain any slashes it could still contain a single `..`

and the slash behavior might change in the future if [this](https://github.com/actix/actix-web/issues/1157) is considered a bug by the actix devs

Reply:
Ah, the repo is checked against the config in the handle_repo() case:
```
    let repo = req.match_info().query("repo");
    let repoconfig = config.get_repoconfig(&repo)?;
```
But yeah, not the id in the handle_build_repo case. I tried it and it cannot contain an escaped slash, but yeah, it could contain a single "..", so best to canonicalize it.

Reply:
Hmm, actually canonicalizing the id from ".." into "" is also a bit weird, it means you can access a different build repo than the one refered to by "id". We don't *currently* do anything with the id other than as a path, but one could imagine doing some kind of permission check against the id, so we should probably also fail if the id element ends up empty.